Windows: Fix raw Dpc offset calculation

dgmcdona opened the pull request:
The original code was still returning this as a pointer that ended up
dereferenced in later steps. However, this pointer value actually needs
to be cast to an `unsigned long long` and decoded first.

dgmcdona committed:
Windows: Fix raw Dpc offset calculation

The original code was still returning this as a pointer that ended up
dereferenced in later steps. However, this pointer value actually needs
to be cast to an `unsigned long long` and decoded first.

dgmcdona commented:
Looking back through the history, I noticed that this originally was an `unsigned long long` in the initial PR but was changed to `pointer` due to concerns about compiler changes and the potential for that type to not exist. Is there some way that I haven’t been able to figure out to access the raw pointer without having it automatically dereference?

ikelos commented:
I think just `int(self.Dpc)` should do it?  Pointers are subclasses on Integers, so the `int` call should just turn it into a number?  You would then lose all the pointerishness of it though...

dgmcdona commented:
I thought so as well, but that doesn't seem to be the case:
```
[ins] In [5]: int(self.Dpc)
Out[5]: 56772929057211

[ins] In [6]: self.Dpc.cast("unsigned long long")
Out[6]: 3104162816094171579

[ins] In [7]: self.Dpc.dereference().vol.offset
Out[7]: 56772929057211
```

ikelos commented:
Errr... Is it getting masked to the address space (because pointers need to fit within the address space), so the high bits are getting masked off?  Oh, yep, that must be [what's happening ](https://github.com/volatilityfoundation/volatility3/blob/369f37ec4366bf5127aac596479b03c5b0819749/volatility3/framework/objects/__init__.py#L411)

```
>>> hex(3104162816094171579)
'0x2b1433a27a44f1bb'
>>> hex(56772929057211)
'0x33a27a44f1bb'
```

Ok, well, the difficulty we have is that clang changed from one name for `unsigned long long`, to another.  I think it only affected linux but it kinda goes to show that reliance on a specific type may one day cause us problems.  The pointer class really should be handling anything that is supposed to be a pointer (I wish OS devs would stop abusing pointers to stash extra data or obfuscate them!).  So, we've got options:

* Stick with the named types we've got and potentially have to special case things if they screw up in the future.
* Write an adaptor that can reread the data at the pointer (and the knowledge of length and endianness from the pointer) to get to the original int - possibly a little convoluted but has the least impact on the existing framework
* Add a method to a pointer to return the original data, although we'd need to be really careful of the name, because it would mask any member of the pointer's subtype.  Probably we'd call it `get_raw_value` or something? 

Which do you think we should go for?  I'm leaning towards one of the last two, but I can't decide which...

dgmcdona committed:
Objects: create `get_raw_value()` method for Pointer

This creates a `get_raw_value()` method for the `Pointer` class that
allows users to access the raw (unmasked) value of a pointer. This was
required in order to decode the encoded `Dpc` pointer that is part of
the `_KTIMER` Windows type. Addition of this type was favored over a
cast to `unsigned long` or `unsigned long long` due to the potential for
future instability of this type due to compiler changes.

See https://github.com/volatilityfoundation/volatility3/issues/1041 for
further discussion around the conversion of `log unsigned int` to
`unsigned long` in `clang`.

See https://github.com/volatilityfoundation/volatility3/pull/1177#discussion_r1650049299
for the original discussion around how to access this pointer in the
`Timers` plugin.

dgmcdona committed:
Extensions: Removes the `get_raw_dpc` method from `KTIMER`

This removes the `get_raw_dpc` method from the `KTIMER` extension class.
This method was inaccurate in that it actually returns the masked
pointer value instead of the full 64-bit value encoded in that member,
which is required in order to correctly decode the 'real' pointer.

The invocation of `get_raw_dpc()` was replaced with
`self.Dpc.get_raw_value()`, which was added in the previous commit.

dgmcdona commented:
I liked option 3 so took a crack at it in 7b9fb916722a0678b666be0f02dd59b13c769951, let me know what you think.

dgmcdona committed:
Framework: Minor version bump

Made an additive change to `Pointer` by adding the `get_raw_value()`
method, so bumping the minor version here. The `get_raw_dpc()` method
was removed from the `KTIMER` extension class, which is currently
unversioned.

ikelos commented on volatility3/framework/objects/__init__.py:
```diff
@@ -410,6 +410,19 @@ def _unmarshall(
+            self.vol.offset, length
+        )
+        struct_format = ("<" if endian == "little" else ">") + formats[length]
+        return struct.unpack(struct_format, raw_data)[0]
```
I think it would be better to use `int.from_bytes` so that it's consistent with the `unmarshall` call above.  You could probably just split the unmarshall above into two functions and expose the `get_raw_value` one...

ikelos reviewed (commented):
Happy with the proposed solution, but we should keep it in sync with the unmarshalling call (and `int.from_bytes` seem better than constructing a format string and then formatting it)...

dgmcdona committed:
Objects: rework new `get_raw_value()` method

Per code review recommendations, splits the `_unmarshall` classmethod
into two components, one of which retrieves the raw value, and the other
that returns the masked pointer. The `get_raw_value` method now calls
the `_get_raw_value` classmethod using its instance information.

dgmcdona committed:
Timers: Adds debug log statement to catch-all exception

dgmcdona committed:
Timers: convert general Exception to InvalidAddressException

ikelos commented on volatility3/framework/symbols/windows/extensions/__init__.py:
```diff
@@ -1268,7 +1254,7 @@ def get_dpc(self):
 
             low_byte = (wait_never) & 0xFF
-            entry = utility.rol(self.get_raw_dpc() ^ wait_never, low_byte)
+            entry = utility.rol(self.Dpc.get_raw_value() ^ wait_never, low_byte)
```
This call now need parameters?

ikelos replied:
Nope, scratch that, sorry, didn't see the additional method.
